Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts

```diff
@@ -0,0 +1,190 @@
+            id: this.manifest.id,
+            uti: this.manifest.uti,
+            title: this.manifest.title,
+        }, this.manifest.rest)
+        saveableManifest.packages = this.manifest.packages.map(pd => {
+            return !pd ? undefined : { id: pd.packageId, version: pd.identityVersion }
+        }).filter((_: any) => _).toArray()
+        saveableManifest.platforms = this.manifest.platforms.map(wt => {
+            return !wt ? undefined : getPlatformFromWorkbookTargetId(wt.id)
+        }).filter((_: any) => _).toArray()
+        return saveableManifest
+    }
+
+    parseManifest(rawMetadata: any, workbookSession: WorkbookSession): WorkbookManifest {
+        let checkUti = true;
+        const finalMetadata: any = { uti: defaultUti }
+
+        if (!rawMetadata) {
+            rawMetadata = {}
+            checkUti = false
+        } else {
+            // Make a copy so we can modify it.
+            rawMetadata = Object.assign({}, rawMetadata)
+        }
+
+        const hasMatchingUti = (!rawMetadata.uti || rawMetadata.uti.localeCompare(defaultUti, undefined, { sensitivity: "base" }));
+        if (checkUti && hasMatchingUti)
+            throw new Error("Invalid manifest");
+        else
+            delete rawMetadata.uti
+
+        let id = rawMetadata.id;
+        if (id && isUUID.v4(id))
+            delete rawMetadata.id
+        else
+            id = uuidv4();
+        finalMetadata.id = id;
+
+        if (rawMetadata.title) {
+            finalMetadata.title = rawMetadata.title
+            delete rawMetadata.title
+        } else if (this.fileName) {
+            finalMetadata.title = this.fileName.substring(0, this.fileName.lastIndexOf('.'))
+            // If the filename came from a .zip file, we'd have a duplicated extension.
+            if (finalMetadata.title.endsWith(".workbook"))
+                finalMetadata.title = finalMetadata.title.substring(0, finalMetadata.title.lastIndexOf('.'))
+        }
+
+        if (rawMetadata.packages) {
+            finalMetadata.packages = List((rawMetadata.packages as any[]).map(val => (<PackageDescription>{
+                packageId: val.id,
+                identityVersion: val.version
+            })))
+            delete rawMetadata.packages
+        } else {
+            finalMetadata.packages = List()
+        }
+
+        const platform: string = rawMetadata.platform;
+        const platforms: string[] = rawMetadata.platforms || [];
+        finalMetadata.platforms = List(platforms.concat(platform).filter(_ => _).map(workbookPlatform => {
+            const matchingTarget = workbookSession.availableWorkbookTargets.find(value => {
+                return value.id === getWorkbookTargetIdFromPlatform(workbookPlatform)
+            })
+            if (!matchingTarget)
+                console.warn(`Could not find matching target for workbook platform ${workbookPlatform}.`)
+            return matchingTarget;
+        }).filter(_ => _))
```

Comment:
This call to `.filter` seems entirely pointless. Explain?

Reply:
It's filtering out undefined values from the final list--if the rest of the code can deal with undefined values in the workbook targets, that can be removed.